vfs: file system abstraction

gebart opened the pull request:
This PR provides a basic virtual file system (VFS) layer for RIOT. The intention is to provide a much simpler layer than what was proposed in #2474 and #1265.

The overall design goals are:
- Provide implementations for all newlib "file" syscalls
- Keep it simple, do not add every possible file operation from Linux VFS.
- Easy to map existing file system implementations for resource constrained systems onto the VFS layer API
- Avoid keeping a central `enum` of all file system drivers that has to be kept up to date with external packages etc.
- Use POSIX `<errno.h>` numbers as negative return codes for errors, avoid the global `errno` variable.
- Only absolute paths to files (no per-process working directory)
- No dynamic memory allocation

The file systems for resource constrained systems that I have been looking at (so far) are [Contiki CFS](https://github.com/contiki-os/contiki/tree/master/core/cfs) and [SPIFFS](https://github.com/pellepl/spiffs).
I have also been looking at Linux kernel headers and books/articles on writing Linux device drivers for more ideas for the design.

The API should be easy to understand for users who are familiar with the POSIX file functions (open, close, read, write, fstat, lseek).

The VFS layer keeps track of mounted file systems and open files, the `vfs_open` function searches the array of mounted file systems and dispatches the call to the file system instance with the longest matching mount point prefix.
Subsequent calls to `vfs_read`, `vfs_write`, etc will do a look up in the table of open files and dispatch the call to the correct file system driver for handling.

`vfs_mount` takes a string containing the mount point, a file system driver specification (`struct file_system`), and an opaque pointer that only the FS driver knows how to use, which can be used to keep driver parameters in order to allow dynamic handling of multiple devices.

Also included is a simple file system implementation of a `static const` file system called ConstFS. This was created as a tool to assist in unit testing of the VFS layer, and as an exercise in implementing file system drivers, and to provide an example for developers who wish to write their own file system driver to hook into the VFS layer.

TODO:
- [x] STDIN, STDOUT, STDERR through vfs_read/write
- [x] Write the newlib syscall wrappers
- [x] Elaborate the design thoughts in the Doxygen documentation
- [ ] Explain clearly how to use the `private_data` members when implementing a file system driver
- [ ] `auto_init` stuff?
- [ ] Shell commands (mount will not be supported due to the dynamic nature of the vfs_mount function)
  - [x] read file
  - [ ] write file
  - [x] list mounted file systems and space utilization (`vfs df`)
  - [x] list files in a directory
- [x] Improve constness to allow for keeping large structures like `vfs_file_ops_t` in flash.
- [x] ConstFS: Implement fstat, lseek, update read to use the file offset
- [x] Mutex for the open files table
- [x] `fcntl` - change flags on already open FDs
- [ ] ~~`ioctl` - communicate with special devices~~ - not needed since kernel is rebuilt every time, easy to use device driver directly
- [x] `stat` (may wrap `open`+`fstat`+`close` sequence as a fall back)
- [x] `statvfs`, `fstatvfs` - information about a mounted file system
- [ ] More unit tests
  - [x] `vfs_open`
  - [x] `vfs_close`
  - [x] `vfs_read`
  - [x] `vfs_write`
  - [x] `vfs_lseek`
  - [x] `vfs_fstat`
  - [x] `vfs_fcntl`
  - [x] `vfs_mount`
  - [x] `vfs_umount`
  - [x] `vfs_rename`
  - [x] `vfs_unlink`
  - [x] `vfs_mkdir`
  - [x] `vfs_rmdir`
  - [x] `vfs_bind`
  - [x] `vfs_opendir`
  - [x] `vfs_closedir`
  - [x] `vfs_readdir`
  - [x] `vfs_normalize_path`
  - [ ] `vfs_iterate_mounts`
  - [x] `vfs_stat`
  - [x] `vfs_statvfs`
  - [x] `vfs_fstatvfs`
  - [x] newlib syscall wrappers
- [x] DevFS device file system
  - [x] basic functionality
  - [x] `auto_init` for mounting
- [ ] DevFS support in drivers
  - [ ] `/dev/uartX` UART devices
  - [x] `/dev/nvramX` NVRAM devices - basic implementation done
  - [ ] `/dev/flashX` Flash devices (requires: Flash API, see #5624, #2239)
  - [ ] `/dev/random` periph/random
- [x] import `xxd`-like dumper from other branch as shell command
- Directory support
  - [x] `opendir`
  - [x] `readdir`
  - [x] `closedir`
  - [x] unit tests for directories
- [x] `vfs_rename`

Not implemented:
- Simultaneous access to the same file.
## How to try it

Right now the only place to run the code is through a newly added embunit unit test batch:

```
make -C tests/unittests tests-vfs term BOARD=native
```

(also successfully tested on BOARD=mulle)
## Open questions
- Does it make sense to add an "inode" layer to separate the file name handling from the file opening?
- Should mode be given as `mode_t` or `int`? (newlib uses int, POSIX uses `mode_t` which is `unsigned long` on 32 bit platforms)
- ~~Should mounted file systems be changed into a linked list? (originally was statically allocated array with fixed maximum mount point length)~~ - implemented

kaspar030 commented:
Nice! Thank you for tackling this. Looks very nice already!

Some thoughts that came up while skimming the diff:

- did you consider extending structs by embedding instead of using pointers, (like netdev2 does)?
- any reason you're using "struct xxx..." within structs, instead of the typedefs?

gebart commented:
@kaspar030 
> - did you consider extending structs by embedding instead of using pointers, (like netdev2 does)?

My thoughts were that some of these structs should be constant and static, e.g. the file system driver, and should not consume valuable RAM, while other parts are variable. Right now I think all of it is kept in RAM, but the goal is to keep the static parts in flash instead and using pointers from the RAM parts to the flash parts.

> - any reason you're using "struct xxx..." within structs, instead of the typedefs?

No particular reason, I started using them in one place to get a pointer to the same type of struct, and then it just kind of kept on. I'll see if it still compiles if I modify it to use the typedefs instead.

Kijewski commented on sys/fs/constfs/constfs.c:
```diff
@@ -0,0 +1,145 @@
+            off += filp->pos;
+            break;
+        case SEEK_END:
+            off += fp->size;
```
break

gebart replied:
addressed

Kijewski commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,382 @@
+    vfs_file_ops_t *f_op;   /**< File operations table */
+    vfs_mount_t *mp;        /**< Pointer to mount table entry */
+    /* TODO: mode, flags, etc */
+    off_t pos;              /**< Current position in the file */
```
I guess using `size_t` would eliminate a bunch of casts in the implementations.

Kijewski commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,382 @@
+     * @param[in]  nbytes   maximum number of bytes to read
+     *
+     * @return number of bytes read on success
+     * @return <0 on error
```
What would `return 0` mean? End of file **OR** try again **OR** query some function which of both?

I'd prohibit the result `0` to make things clear. Return `-ENODATA` for EOF, and `-EAGAIN` for try again.

I am not sure about `ENODATA`, though. Maybe `ERANGE`?

gebart replied:
`return 0` would mean that there is no data to read, either because the position is at or past the end of file, or in the case of a device, that nothing is available to read right now. EOF is not necessarily a permanent condition, it depends on if the file is later written to extend it. In the case of a device, for example UART, the 0 read means that nothing has been received, but another read may give something back.
**To summarise**, return 0 means try again later or give up, it depends on the application, and it may not be possible for the system developers to know in advance which of the two alternatives is the right choice for any given occasion.

The POSIX read function also does not set an error for EOF, it just returns 0. http://pubs.opengroup.org/onlinepubs/9699919799/functions/read.html

The glibc manual pages on low level I/O also explicitly states

>A value of zero indicates end-of-file (except if the value of the size argument is also zero). This is not considered an error. If you keep calling read while at end-of-file, it will keep returning zero and doing nothing else.

https://www.gnu.org/software/libc/manual/html_node/I_002fO-Primitives.html#I_002fO-Primitives

gebart commented:
updated todo-list with some more ideas.

vincent-d commented:
This looks very nice.

I was working on the same topic, but my work is much less advanced than yours. I'm trying to get SPIFFS working and we worked on a low level flash interface as well. I'll push some work soon and will try to rebase my SPIFFS port on top of your vfs.

Thanks for that work

gebart commented:
@vincent-d Nice! I have had plans to add SPIFFS but I did not feel that I knew where to begin, so I started with this instead, it's good to see that others are working towards the same goal.
I am currently under way to add a device file system, DevFS, similar to /dev on Linux, in order to be able to access low level devices from the shell, such as NVRAM. Maybe we can fit the generic flash driver onto that as well? (see TODO in top post)

gebart commented:
squashed, WIP commit list was getting out of hand.

kaspar030 commented:
How do we deal with directories?

gebart commented:
Updated with more work on DevFS, including:

 - DevFS `auto_init` mounts `/dev`
 - Shell command `vfs` to access the VFS mounted file systems from the shell
 - Newlib fixes
 - basic `fcntl` implementation
 - VFS file operations implementation for `/drivers/nvram` to access the contents of NVRAM devices via VFS.
 - Configured NVRAM VFS on DevFS for Mulle (`/dev/mulle-fram`).
 - improved unit tests

gebart commented:
@kaspar030 not sure. CFS and SPIFFS both implement specific `opendir`, `closedir`, `readdir`. Maybe those functions should be added as part of `vfs_file_system_ops_t`, what do you think?

vincent-d commented on sys/vfs/vfs.c:
```diff
@@ -0,0 +1,660 @@
+        return -ENOENT;
+    }
+    /* Increment open files counter for this mount */
+    atomic_inc(&_vfs_mounts[md].open_files);
```
This should not be done here but only in vfs_open, otherwise the unlink also increments the counter and umount fails afterwards.

gebart replied:
Actually, this is on purpose as a kind of semaphore to avoid races between umount and open/unlink. There were a few missing decrements of open_files in vfs_unlink though that I have fixed now.

vincent-d commented on sys/vfs/vfs.c:
```diff
@@ -0,0 +1,660 @@
+            /* path name is shorter than the mount point name */
+            continue;
+        }
+        else if (name[len] != '/' && name[len] != '\0') {
```
len > 1 should be tested first, otherwise it does not work with "/" as a mountpoint

gebart replied:
right, good catch!

gebart replied:
addressed

vincent-d commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,453 @@
+    int flags;              /**< File flags */
+    off_t pos;              /**< Current position in the file */
+    kernel_pid_t pid;       /**< PID of the process that opened the file */
+    void *private_data;     /**< File system driver private data, implementation defined */
```
I propose we add a private int value as well, either alongside or in an union with the void *private_data.
It would become handy to store a file handle returned by the underlying file system implementation, for instance.

gebart replied:
I'll change it to `union { void *, int }`. I don't think we should waste RAM for the esoteric case of simultaneous int+pointer, (that special case can be worked around by pointing to a secondary `struct { void *, int }` any way.)

vincent-d commented on sys/vfs/vfs.c:
```diff
@@ -0,0 +1,663 @@
+                _free_mount(md);
+                return -EINVAL;
+            }
+            if (mountp->mount_point[len - 1] == '/') {
```
len should be > 1 here also

gebart replied:
addressed

gebart commented:
I'll do an update tomorrow morning with opendir, readdir, closedir support for devfs and constfs

gebart commented:
Updated with the directory interface. No unit tests yet or ConstFS, DevFS support for directories.

miri64 commented:
I know FAT is not the most ideal FS for our purposes with all its layers of backwards compatibility, but we also should also consider porting the [FatFS library](http://elm-chan.org/fsw/ff/00index_e.html) as a package using this interface. After all: FAT is still the file system type most SD cards are shipped with. I saw this library used on systems that are way more constraint than the ones RIOT targets (the [CCC's https://github.com/r0ket/r0ket](r0ket badge) e.g.), so space shouldn't be an issue.

kaspar030 commented:
> we also should also consider porting the FatFS library as a package using this interface.

+1

I was actually thinking about adding support in order to get a feeling for vfs.

gebart commented:
@kaspar030 I think the API is stable enough to try to implement some file systems on top of now.
IMO we need to do a few example implementations to see if there are any opportunities to improve the API before merging this PR.

miri64 commented:
> I was actually thinking about adding support in order to get a feeling for vfs.

+1

> IMO we need to do a few example implementations to see if there are any opportunities to improve the API before merging this PR.

+1

vincent-d commented:
I already try my hand at implementing SPIFFS (see #5625) and the API looks already pretty good to me.
I also agree we should try more implementations (Coffee, FatFS) but what will miss now is the low level interface.
But we also have a solution :D (see #5624).

I'm really glad this topic is moving forward.

gebart commented:
I just realized a mistake in the `readdir` declaration. I will change it so that `return 1` means `entry` was updated, `return 0` means end of stream.

gebart commented:
Updated the PR with a basic `ls` command. Nested mounts are not displayed within `ls` output.

gebart commented:
doing a little build test on Murdock.
(Murdock was idle according to the status page)

kaspar030 commented:
Should probably be done as follow-up, but it would be nice if either per fs or per file, there's be an execute-in-place flag if files are in CPU-accessible flash and continuous.

gebart commented:
@kaspar030 good idea, but yeah, let's leave that as a follow up

vincent-d commented:
@gebart Did you think about adding support for renaming files ? Other usefull functions could also be ftell() and feof(). Waht do you think ?

gebart commented:
ftell and feof are part of the higher level libc components, which should work on top of the POSIX layer. If you need to know the position in the file you can always do an `pos = lseek(fd, 0, SEEK_CUR)` and to see if you are at EOF you check the return value of read to see if it is zero.
I didn't test but I guess that the newlib fopen, fclose, fread, ftell etc should "just work" on top of this, with the syscall implementations (won't work on native though because `open` etc are still the regular glibc versions which operate in the Linux user space outside of RIOT)

gebart commented:
@vincent-d but rename is a useful addition. I think we need to constrain it to rename between files on a single mount point for now, for easier implementation.

I propose the following:

global function:
```
int vfs_rename(const char *old, const char *new);
```

file system driver function (in `vfs_file_system_ops_t`):
```
int (*rename) (vfs_mount_t *mountp, const char *old, const char *new);
```

vincent-d commented:
> I think we need to constrain it to rename between files on a single mount point for now, for easier implementation.

Agreed

@gebart I agree with you proposal

gebart commented:
I have pushed a proposal for `vfs_rename`, and rebased on the updated #5622.

vincent-d commented on sys/include/vfs.h:
```diff
@@ -65,6 +65,46 @@ extern "C" {
+ * compiler error):
+ *
+ * @attention @code
+ * #if VFS_DIR_BUFFER_SIZE < (sizeof(my_DIR))
```
@gebart I'm not sure this can actually work. Is it really possible to use sizeof in a preprocessor expression ?

gebart replied:
Right, I'll change the example to a literal number instead
On Jul 13, 2016 6:09 PM, "Vincent Dupont" <notifications@github.com> wrote:

> In sys/include/vfs.h
> <https://github.com/RIOT-OS/RIOT/pull/5616#discussion_r70656559>:
>
> > + * systems where a single pointer is not enough space for its directory stream
> > + * state, e.g. SPIFFS.
> > + *
> > + * Guidelines:
> > + *
> > + * SPIFFS requires a sizeof(spiffs_DIR) (6-16 bytes, depending on target
> > + * platform and configuration) buffer for its DIR struct.
> > + *
> > + * @attention File system developers: If your file system requires a buffer for
> > + * DIR streams that is larger than a single pointer or @c int variable, ensure
> > + * that you have a preprocessor check in your header file (so that it is
> > + * impossible to attempt to mount the file system without running into a
> > + * compiler error):
> > + *
> > + * @attention @code
> > + * #if VFS_DIR_BUFFER_SIZE < (sizeof(my_DIR))
>
> @gebart <https://github.com/gebart> I'm not sure this can actually work.
> Is it really possible to use sizeof in a preprocessor expression ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5616/files/7df540db96c9b48bfb73e50583a7c1be812dac0e..55a9475dd05d0496107fe0811943ae52d91aa5d6#r70656559>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AATYQvjO9OZIJqXg5wckHpATesFyo3Rzks5qVQ3CgaJpZM4JIYEa>
> .
>

gebart replied:
addressed

gebart commented:
Q: Should the implementation of the list of currently mounted file systems be modified to a linked list where the nodes are allocated by the user mounting the FS?

Currently it is implemented as a fixed size array statically allocated inside VFS with fixed size statically allocated string buffers for the mount point names, and the names given to `vfs_mount` are copied into the array buffers.

RIOT-notifier commented:
> Q: Should the implementation of the list of currently mounted file
systems be modified to a linked list where the nodes are allocated by the
user mounting the FS?

I think this is the more flexible solution that gives at least some control
to the user, while not requiring any kind of dynamic allocation.

miri64 commented:
(that was me)

gebart commented:
so change vfs_mount_t to:
```c
struct vfs_mount_struct; /* not vfs_mount because of name collision with the function */
typedef struct vfs_mount_struct vfs_mount_t;
struct {
    vfs_mount_t *next;
    const vfs_file_system_t *fs; /**< The file system driver for the mount point */
    const char *mount_point;     /**< Mount point, e.g. "/mnt/cdrom" */
    atomic_int_t open_files;     /**< Number of currently open files */
    void *private_data;          /**< File system driver private data, implementation defined */
};
```
One thing that won't be enforced is the normalization of the mount point path.

miri64 commented:
> One thing that won't be enforced is the normalization of the mount point path, unless mount_point is made non-`const`.

What would that imply?

gebart commented:
@miri64 That would imply that `vfs_mount("blah//hej/./asdf")` would actually mount at that exact path, and not "/blah/hej/asdf", which means it will not be possible to access files there if the tools to access the files use path normalization, like the shell commands `ls` and `vfs` do.

gebart commented:
we could do a check for it using a temporary buffer, but we can't modify it since it is const and we aren't allowed to dynamically allocate a string.

miri64 commented:
What implication would removing the `const` modifier have?

gebart commented:
@miri64 removing the `const` modifier would require the mount point string to be located in RAM instead of flash. RAM is rare on our target platforms so we should try to conserve it as best possible.

gebart commented:
Updated with a linked list implementation for `vfs_mount`
This requires changes to the user code mounting the file systems to provide a `vfs_mount_t` struct with at least `mount_point` and `fs` set to something sensible.

gebart commented:
rebased on latest master

kaspar030 commented:
> Updated with a linked list implementation for vfs_mount

Wouldn't it make sense to re-use the available linked list code? Maybe base it on #5629? (follow-up?)

gebart commented:
@kaspar030 right, I was lazy. But right now I'll put the update on hold until #5629 is merged.

gebart commented:
@kaspar030 Actually, after looking at the code for a little bit, I am still not sure on how to use clist as a normal singly linked list. What should my data structures look like in order to create a list of `vfs_mount_t` structs?
How do I traverse it and `strcmp` the names at each step like in `_find_mount`?

kaspar030 commented:
(this is based on the changes in #5629)
@gebart Instead of the next member of vfs_mount_t there needs to be a 'clist_node_t list_entry' (name doesn't matter, not a pointer).
Then there needs to be one 'clist_node_t' that will be used as list handle. Let's call it '_vfs_mount_list'.

The clist_*() functions now take &_vfs_mounts_list as list parameter and a pointer to the list_entry field as object parameters.
Adding elements to front/back can then be done with 'clist_lpush()/clist_rpush(&_vfs_mounts_list, &mount->list_entry);

Traversal can be done like this:

         clist_node_t *node = _vfs_mounts_list.next;
         if (! node) {
             // list empty
         }

         do {
             node = node->next;
             // do something with node
         } while (node != _vfs_mounts_list.next);

gebart commented:
@kaspar030 I updated `vfs_mount`, `vfs_umount`, `_find_mount` to use clist.h instead of hand crafted list handling. It passes all the unit tests so I guess it works, but could you take a quick look at those functions to see that I am using it correctly?

kaspar030 commented:
> could you take a quick look at those functions to see that I am using it correctly?

Looks perfect. But you not feeling confident about it's usage makes me wonder if the API is intuitive like this?

kaspar030 commented:
> But you not feeling confident about it's usage makes me wonder if the API is intuitive like this?

Didn't see your comment on the clist PR.

gebart commented:
AVR and MSP430 are failing with missing `<fcntl.h>` and `<sys/stat.h>`. MSP430 would work if it supported newlib (#4766 abandoned WIP, available for adoption). For AVR I don't know, should we provide the basic headers?

gebart commented:
- Added `statvfs`, `stat`, `fstatvfs` implementations.
 - Added `vfs df` shell command

gebart commented:
Rebased after #5629 was merged.

kaspar030 commented:
I get this when compiling for native (default example with added ```USEMODULE += vfs```):

```
"make" -C /home/kaspar/src/riot.tmp/sys/shell/commands
sc_vfs.c: In function ‘_errno_string’:
sc_vfs.c:64:17: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         if (res <= buflen) {
                 ^~
sc_vfs.c: In function ‘_read_handler’:
sc_vfs.c:175:22: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         else if (res > line_len) {
                      ^
sc_vfs.c:186:30: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (size_t k = 0; k < res; ++k) {
                              ^
sc_vfs.c:194:30: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (size_t k = 0; k < res; ++k) {
                              ^
cc1: all warnings being treated as errors
```

Probably those size_t need to be changed to ssize_t?

kaspar030 commented:
On samr21-xpro with the arch linux toolchain:

```
In file included from /home/kaspar/src/riot.tmp/sys/newlib/syscalls.c:42:0:
/home/kaspar/src/riot.tmp/sys/include/vfs.h:34:50: fatal error: sys/statvfs.h: No such file or directory
 #include <sys/statvfs.h> /* for struct statvfs */
                                                  ^
compilation terminated.
```

(no sys/statvfs.h in the newlib includes present)

gebart commented:
@kaspar030 I have prepared a statvfs.h header for use with VFS, I'll post it tomorrow, probably. Thanks for testing.

kaspar030 commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,796 @@
+#define VFS_ANY_FD (-1)
+
+/* Forward declarations */
+struct vfs_file_ops;
```
I think you can omit these empty struct definitions.

kaspar030 commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,796 @@
+ */
+typedef struct vfs_file_ops vfs_file_ops_t;
+
+struct vfs_dir_ops;
```
here

kaspar030 commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,796 @@
+ */
+typedef struct vfs_dir_ops vfs_dir_ops_t;
+
+struct vfs_file_system_ops;
```
here

kaspar030 commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,796 @@
+ */
+typedef struct vfs_file_system_ops vfs_file_system_ops_t;
+
+struct vfs_mount_struct; /* not vfs_mount because of name collision with the function */
```
here

kaspar030 commented:
@gebart Do you have a tool for creating constfs datastructures from a given directory (at compile time)? Otherwise I start hacking one.

kaspar030 commented:
Well, it's a little hacked, but works:

https://github.com/kaspar030/RIOT/tree/vfs/dist/tools/mkconstfs

If you deem it usable, feel free to cherry-pick the commit.

kaspar030 commented:
When I have only e.g., ```/test``` as mountpoint, ```ls /``` doesn't show it.
So probably when doing readdir, the mountpoints are not considered.
It's probably ugly to implement, but it might be needed (expected by users)?

kaspar030 commented:
I'm collecting my fixes in https://github.com/gebart/RIOT/pull/9.

gebart commented:
@kaspar030 
>When I have only e.g., ```/test``` as mountpoint, ```ls /``` doesn't show it.
So probably when doing readdir, the mountpoints are not considered.
It's probably ugly to implement, but it might be needed (expected by users)? 

Yes, it's ugly to implement. I have thought about this too.

Another issue is that the returned paths from readdir are VFS mount-point-relative, not VFS absolute.

Do you have any ideas for solutions that do not involve implementing inodes for every VFS file? (memory constraint)

gebart commented:
We could make vfs_readdir first return all mount points matching `path` before returning anything from the actual file system, but it's kind of a mess

gebart commented:
Is it OK if I squash this now? The commit list is getting very long again..

kaspar030 commented:
> Is it OK if I squash this now?

Sure, squash!

gebart commented:
rebased, squashed to reduce the length of the commit list.

 - Added posix compatible headers for `sys/types.h`, `sys/stat.h`, `sys/statvfs.h`, `fcntl.h`

gebart commented:
@kaspar030 would you like to try to implement something for readdir so that it can return mount points as well?

gebart commented:
Added `vfs cp`, `vfs rm`, `vfs mv` shell commands

gebart commented:
fixed a bug in readdir for devfs (I was clumsy when updating devfs to use clist)

gebart commented:
I think this PR is ready for a review now.

- added unit tests of all vfs_ stubs when a file system operation is not supported by the driver.
- Fixed a null pointer dereference when calling vfs_closedir twice

I don't think it is feasible to write unit tests for all operations in all file systems as a part of this PR, the current testing level does testing of the most common functionality with some spot checks of the more uncommon cases.
It might be feasible to write more unit tests when we have a real file system to test against (e.g. #5625)

gebart commented:
Added some more text in the Doxygen VFS module documentation

gebart commented:
I would like to get the ball rolling on this one to avoid getting caught up in the feature freeze and release preparations. @kaspar030 do you have the time to review this soon-ish?

kaspar030 commented:
> I would like to get the ball rolling on this one to avoid getting caught up in the feature freeze and release preparations. @kaspar030 do you have the time to review this soon-ish?

Yeah, let's get this in. I'll review now.

kaspar030 commented on sys/newlib/syscalls.c:
```diff
@@ -39,6 +39,7 @@
 #include "reboot.h"
 #include "irq.h"
 #include "log.h"
+#include "vfs.h"
```
This needs to be ```#ifdef MODULE_VFS```, otherwise I get this:

```
"make" -C /home/kaspar/src/riot.tmp/sys/newlib
In file included from /home/kaspar/src/riot.tmp/sys/newlib/syscalls.c:42:0:
/home/kaspar/src/riot.tmp/sys/include/vfs.h:59:50: fatal error: sys/statvfs.h: No such file or directory
 #include <sys/statvfs.h> /* for struct statvfs */
```

kaspar030 commented on sys/include/vfs.h:
```diff
@@ -0,0 +1,818 @@
+ * @return fd number on success (>= 0)
+ * @return <0 on error
+ */
+int vfs_bind(int fd, int flags, vfs_file_ops_t *f_op, void *private_data);
```
Is there a reason why ```f_op``` is not const?

gebart replied:
legacy, f_op in the struct used to be non-const before. Addressed now.

kaspar030 commented on sys/fs/constfs/constfs.c:
```diff
@@ -0,0 +1,304 @@
+#include "fs/constfs.h"
+#include "vfs.h"
+
+#define ENABLE_DEBUG (1)
```
These need to be disabled for merging.

kaspar030 commented:
While testing, I realized that the posix functions (open, read, write) don't work at all on native. Should we maybe disable vfs completely for native?

kaspar030 commented:
> don't work at all on native.

supplying wrappers seems to work. I'll come up with sth.

kaspar030 commented:
> supplying wrappers seems to work. I'll come up with sth.

See https://github.com/gebart/RIOT/pull/11.

gebart commented:
- Merged gebart#11
 - Enable unit testing of syscall wrappers for native and newlib (was: only newlib)

cgundogan commented on cpu/atmega_common/include/sys/types.h:
```diff
@@ -7,22 +7,49 @@
 extern "C" {
 #endif
 
-typedef int16_t suseconds_t;
```
it's probably a good idea to move the proposed atmega changes to a seperate PR, as they are mostly unrelated to this new feature. (here and above)

gebart commented:
Agreed, there is some header juggling that needs to be resolved before
Murdock will agree with this PR. I'll prepare a separate PR when I have
some time to fix it.

On Sep 21, 2016 09:17, "Cenk Gündoğan" <notifications@github.com> wrote:

> *@cgundogan* commented on this pull request.
> ------------------------------
>
> In cpu/atmega_common/include/sys/types.h
> <https://github.com/RIOT-OS/RIOT/pull/5616#pullrequestreview-894582>:
>
> >  #include <stddef.h>
>
>  #ifdef __cplusplus
>  extern "C" {
>  #endif
>
> -typedef int16_t suseconds_t;
>
> it's probably a good idea to move the proposed atmega changes to a
> seperate PR, as they are mostly unrelated to this new feature. (here and
> above)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5616#pullrequestreview-894582>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQnMKf0BZ3fq4EWUvTrWFRKj9VrMRks5qsNoggaJpZM4JIYEa>
> .
>

miri64 commented:
@gebart do you plan to harmonize your file descriptor table with https://github.com/RIOT-OS/RIOT/blob/master/sys/posix/fd.c somehow?

miri64 commented:
(I'm asking primarily to for my POSIX socket sock wrapper ;-)).

gebart commented:
This has been put on hold temporarily because of some important deadlines at work taking a lot of time from me. I'll try to clean up the header situation and I will look over the posix fds when I have a chance to get back to this.

miri64 commented:
Postponed due to Feature Freeze.

vincent-d commented:
@gebart any chances you can work on this soon in the next weeks ? 

It seems also that this PR could me merged almost as is, then new features or improvement could come later on.
This is a great work and it would be nice if we could also add the mtd and spiffs support soon. We can help on some task if needed.

gebart commented:
I hope to clean this up within the next week or so. I'd like to see this merged even though I don't have much time to work on RIOT related stuff at the moment.
I have rebased on latest `master`, but I think there were some problems with missing or conflicting C library headers for some platform(s) last time I worked on this.
newlib and the other libc's provide different levels of support for the functions and types used by this PR, especially the POSIX part.
@vincent-d if you want to help, you could try to get the buildtests to pass for all boards when building using the Docker container (`make buildtest BUILD_IN_DOCKER=1`) on all examples and tests.
I think there may be some header issues with especially the msp430 and avr platforms.

kaspar030 commented:
@gebart do you currently need murdock's results? I'm maintaining it ATM but can't without stopping builds...

gebart commented:
@kaspar030 I was just being lazy and waiting for Murdock to run a full build for me. I'm on a laptop so it'll take a while to get the same results without Murdock. I think I've fixed all of the libc header issues, with a bit of help from @vincent-d.

gebart commented:
@kaspar030 but it's OK to shut it down if you are working on Murdock.

gebart commented:
@miri64 
I was looking over the posix socket fd parts, comparing between vfs and fd.h. In the current state, would we gain anything from unifying file FDs and socket FDs?
From my point of view, we would have to add another level of indirection via an array which would hold pointers to vfs_file_t objects or socket_t objects. Merging socket_t and vfs_file_t would waste memory unless we made parts of the resulting object into a transparent union, with some added user confusion when working with the objects..

Maybe we can leave the unification of FDs as an open issue until there is any obvious benefit from doing it? I would like to see this merged before the 2017.01 release, if possible.

kYc0o commented:
Do we have news on this?

kaspar030 commented:
> Maybe we can leave the unification of FDs as an open issue until there is any obvious benefit from doing it? I would like to see this merged before the 2017.01 release, if possible.

+1

miri64 commented:
> > Maybe we can leave the unification of FDs as an open issue until there is any obvious benefit from doing it? I would like to see this merged before the 2017.01 release, if possible.
>
> +1

I'm also fine with that.

LudwigKnuepfer commented:
> Maybe we can leave the unification of FDs as an open issue until there is any obvious benefit from doing it? I would like to see this merged before the 2017.01 release, if possible.

👍

gebart commented:
rebased, squashed minor Doxygen and cppcheck fixes into primary vfs commit.

miri64 commented:
Mh... in current https://ci.riot-labs.de/ this PR is marked as successful build, but here it is still marked as pending. @kaspar030 any hints?

kaspar030 commented:
Green! I'll do another round of tests tomorrow. Looking forward to have this merged!

miri64 commented:
@kaspar030 did you?

vincent-d commented on sys/posix/include/fcntl.h:
```diff
@@ -0,0 +1,75 @@
+
+/* Constants used for the flags in fcntl and open */
+/* The names of the constants are defined by POSIX but most of these numbers
+ * were chosen somewhat arbitrarily */
```
When using fopen on cortex_m, the flags value seems mixed up. fopen does not set the same flags value than these ones.

gebart replied:
right, newlib provides fcntl.h, we should only use our own if there's not a libc-provided fcntl.h. Will update

gebart replied:
Added an `#if` block to include the newlib-provided fcntl.h if building against newlib. Could you test it and see if that fixes your issues?
Especially the O_CREAT, O_TRUNC, O_EXCL flags were different than what newlib has defined.

vincent-d replied:
Looks OK

gebart commented:
rebased

gebart commented:
Added condition to use system fcntl.h if MODULE_NEWLIB is enabled.

kaspar030 commented:
Weird Murdock output. Let's retry.

kaspar030 commented:
For the record:
```
Errors:

unittests:remote-pa:
Traceback (most recent call last):
  File "./dist/tools/compile_test/compile_test.py", line 189, in 
    print_output_for(errors, 'Errors', Termcolor.red)
  File "./dist/tools/compile_test/compile_test.py", line 145, in print_output_for
    print('%s'  % output.getvalue())
UnicodeEncodeError: 'ascii' codec can't encode character u'\ufffd' in position 12071: ordinal not in range(128)
```

vincent-d commented on sys/shell/commands/shell_commands.c:
```diff
@@ -133,6 +133,11 @@ extern int _ccnl_fib(int argc, char **argv);
+extern int _ls_handler(int argc, char **argv);
+#endif
+
 const shell_command_t _shell_command_list[] = {
```
vfs and ls commands are not in the `_shell_command_list`

gebart replied:
oops, that's probably a rebase mistake, I'll fix

kaspar030 commented:
[Here](https://ci.riot-labs.de/devel/RIOT-OS/RIOT/5616/60a481483b6716c59cbeeca1b00042fa23bf66fd/output.html#error0) you can see some unittests overflow ram.

gebart commented:
Hilarious, I had to add a work around the output recording in compile_test
to get it to replace the invalid characters with 0xfffd, and now something
else broke because of it.

On Jan 20, 2017 15:19, "Kaspar Schleiser" <notifications@github.com> wrote:

> For the record:
>
> Errors:
>
> unittests:remote-pa:
> Traceback (most recent call last):
>   File "./dist/tools/compile_test/compile_test.py", line 189, in
>     print_output_for(errors, 'Errors', Termcolor.red)
>   File "./dist/tools/compile_test/compile_test.py", line 145, in print_output_for
>     print('%s'  % output.getvalue())
> UnicodeEncodeError: 'ascii' codec can't encode character u'\ufffd' in position 12071: ordinal not in range(128)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5616#issuecomment-274082511>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQuneT3J-X1hNom4QJeeHf5GaVeOVks5rUMJ0gaJpZM4JIYEa>
> .
>

miri64 commented:
> Hilarious, I had to add a work around the output recording in compile_test
to get it to replace the invalid characters with 0xfffd, and now something
else broke because of it.

Computers, they are such a joy.. :rofl:

gebart commented:
ping @vincent-d @kaspar030 @Kijewski 
rebased and waiting for Murdock.

gebart commented on sys/fs/devfs/auto_init_devfs.c:
```diff
@@ -0,0 +1,37 @@
+ */
+
+/**
+ * @ingroup     auto_init_saul
```
addressed

gebart commented:
Updated the Doxygen blocks in some of the files where there was some copy pasta. Rebased

gebart commented:
@vincent-d OK to squash the Doxygen changes?

vincent-d commented:
> @vincent-d OK to squash the Doxygen changes?

Yes, looks good to me

gebart commented:
squashed

gebart commented:
Rebased after #5688 was merged, updated vfs code accordingly.

kaspar030 commented:
There's some trouble with the new atomics. I guess on msp430, they're not used enough to be spotted before. But why does tests_cpp11_condition_variable suddenly break?

gebart commented:
Yes, it's related to the comma operator in stdatomic when used with __sync
builtins

On Feb 10, 2017 10:06 AM, "Kaspar Schleiser" <notifications@github.com>
wrote:

> There's some trouble with the new atomics. I guess on msp430, they're not
> used enough to be spotted before. But why does
> tests_cpp11_condition_variable suddenly break?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5616#issuecomment-278892710>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQmwD8xlF2wgAVbP9LfK7vyu3wsPbks5rbCimgaJpZM4JIYEa>
> .
>

gebart commented:
rebased on top of #6587, it should fix the msp430 problems in vfs.c

kaspar030 commented:
Looks like the CI's are ceasing to complain! I'd say, please squash.

gebart commented:
There are some new commits regarding the mips platform which need further
review and should probably be split into its own PR.
The reason for the mips patch is that statvfs.h needs to know when we are
building with newlib in order to provide some additional typedefs which
were only recently added to newlib. The newlib in mips-mti-elf is based on
upstream sources from before 2014.

On Feb 22, 2017 5:48 PM, "Kaspar Schleiser" <notifications@github.com>
wrote:

> Looks like the CI's are ceasing to complain! I'd say, please squash.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5616#issuecomment-281728111>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQlN8ixLrCgOEoDLqg5oqUCBQ-1khks5rfGb1gaJpZM4JIYEa>
> .
>

gebart commented:
The current empty syscalls implementation on MIPS only means that VFS will not work with the POSIX wrapper layer, but it will at least compile. All VFS use on MIPS will therefore only work through the dedicated `vfs_xxx` functions, the POSIX calls will be passed on to the semihosting environment, with unknown results.
I don't have any boards to test on but I assume that the unit tests for the POSIX wrapper in tests-vfs will fail on this architecture, but all other tests should succeed.

gebart commented:
Rebased on top of #6587. Squashed and ready for review.

miri64 commented:
Does this need a re-ACK?

gebart commented:
@miri64 Needs at least someone to approve the workaround for the mips board in the beginning of fcntl.h and statvfs.h

gebart commented:
@vincent-d Does your ACK hold?

vincent-d commented:
Yes! ACK

miri64 commented:
qemu error on Jarvis seems unrelated and Murdock (1) is still the imperative CI, so merge at will.

gebart commented:
Yay! :D

miri64 commented:
Congratz \o/

vincent-d commented:
Great!!

kaspar030 commented:
Congrats and a lot of thanks to @gebart! Good job!

emmanuelsearch commented:
great!!

kYc0o commented:
Super! Thanks @gebart !
